File: solutions/day2_1.py
with open('/Users/vincent/Documents/AoC23/inputs/input2.txt', 'r') as f:
    data = [l.strip() for l in f.readlines()]

lines = [l[l.index(": ")+1:].strip() for l in data]

max_cubes = {
    "red":12,
    "green":13,
    "blue":14
}
valid_line_sum = 0
impossible = []
for idx in range(len(lines)):
    l = lines[idx]
    handfuls = l.split(';')
    valid = True
    for h in handfuls:
        if not valid:
            break
        cubes = [c.strip() for c in h.split(',')]
        for c in cubes:
            num = int(c.split(' ')[0])
            color = c.split(' ')[1]
            if num > max_cubes[color]:
                impossible.append(idx + 1)
                valid = False
                break

    if valid:
        valid_line_sum += idx + 1
print(valid_line_sum)
print(impossible)

File: solutions/day2_1.mjs
import * as fs from "fs";
const data = fs
    .readFileSync("/Users/vincent/Documents/AoC23/inputs/input2.txt", { encoding: "utf-8", flag: "r" })
    .trim()
    .split("\n")
    .map((line) => line.split(": "))
    .map((line) => [
        parseInt(line[0].split(" ")[1]),
        line[1]
            .split(";")
            .map((l) =>
                l
                    .split(", ")
                    .map((l) => {
                        let t = l.trim().split(" ");
                        return { [t[1]]: parseInt(t[0]) };
                    })
                    .reduce((acc, cur) => ({ ...acc, ...cur }), {})
            )
            .map((l) => {
                if (l["red"]) {
                    l["red"] -= 12;
                }
                if (l["green"]) {
                    l["green"] -= 13;
                }
                if (l["blue"]) {
                    l["blue"] -= 14;
                }
                return l;
            })
            .map((l) => Object.values(l).map((l) => l <= 0))
            .flat()
            .reduce((acc, cur) => acc && cur, true),
    ])
    .filter((line) => line[1])
    .map((line) => line[0])
    .reduce((acc, cur) => acc + cur, 0);

console.log(data);
